### PR TITLE
Pass directory path from call-site

### PR DESCRIPTION
**Pull request by am11**
When init-compiler is sourced (as opposed to `eval`'d), the script was giving a non-fatal error, e.g. on mono x64: `pipeline-logging-functions.sh: No such file or directory`. We use Azure logging functions in error cases, so it went unnoticed.

This PR uses the same simple mechanism for sourced scripts we have used in other places; pass the directory path from all call-sites as an argument. Also quoted the positional arguments.

### CODE REVIEW

**Comment by am11** [comment]
cc @jkoritzinsky, @janvorli, noticed thi when I was looking at mono logs for another issue: https://dev.azure.com/dnceng/public/_build/results?buildId=1397492&view=logs&j=38c63b74-c97f-5bda-376d-1084248ecb57&t=b7130259-7da2-5518-9023-82792e689970&l=98

File: eng/common/native/init-compiler.sh [hunk 1]
```diff
@@ -5,21 +5,23 @@
 
 if [[ "$#" -lt 2 ]]; then
```
Comment (am11) [inline eng/common/native/init-compiler.sh:6]:
```suggestion
if [[ "$#" -lt 3 ]]; then
```

**Comment by am11** [comment]
[Before](https://dev.azure.com/dnceng/public/_build/results?buildId=1397492&view=logs&j=38c63b74-c97f-5bda-376d-1084248ecb57&t=b7130259-7da2-5518-9023-82792e689970&l=98):
>   Running 'bash -c 'source /__w/1/s/eng/common/native/init-compiler.sh x64 clang &&  ....
>
>  /__w/1/s/eng/common/native/init-compiler.sh: line 16: /__w/1/s/artifacts/obj/mono/Linux.x64.Debug/../pipeline-logging-functions.sh: No such file or directory

(the build still continues and succeeds because it's not running with `set -e`)

[After](https://dev.azure.com/dnceng/public/_build/results?buildId=1397798&view=logs&j=38c63b74-c97f-5bda-376d-1084248ecb57&t=b7130259-7da2-5518-9023-82792e689970&l=93):
>  Running 'bash -c 'source /__w/1/s/eng/common/native/init-compiler.sh "/__w/1/s/eng/common/native" "x64" "clang" &&  ....

It regressed in #59018.

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Comment by am11** [comment]
win x64 failure is unrelated to PR changes: #29683.

**Pull request merged**